### PR TITLE
Update Cargo, build curl/OpenSSL statically via features

### PR DESCRIPTION
**Pull request by alexcrichton**
In addition to to updating Cargo's submodule and Cargo's dependencies,
this also updates Cargo's build to build OpenSSL statically into Cargo
as well as libcurl unconditionally. This removes OpenSSL build logic
from the bootstrap code, and otherwise requests that even on OSX we
build curl statically.

### CODE REVIEW

File: src/bootstrap/dist.rs [hunk 1]
```diff
@@ -990,12 +989,12 @@ impl Step for PlainSourceTarball {
                    .arg("--debug")
                    .arg("--vers").arg(CARGO_VENDOR_VERSION)
                    .arg("cargo-vendor");
-                if let Some(dir) = builder.openssl_install_dir(builder.config.build) {
-                    builder.ensure(native::Openssl {
-                        target: builder.config.build,
-                    });
-                    cmd.env("OPENSSL_DIR", dir);
-                }
+                // if let Some(dir) = builder.openssl_install_dir(builder.config.build) {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/dist.rs]: Hm, I think we probably want to keep this? Or I guess maybe just delete it since we're not managing openssl on our self?

**Review by Mark-Simulacrum — COMMENTED** [review]
Looks generally good to me -- left a few nits.

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (alexcrichton) [reply src/bootstrap/dist.rs]: Oops I meant to delete this! I built OpenSSL in at least one of our containers so I'm hoping that this isn't necessary any more, but if it fails on the bots I'll probably need to find a workaround

**Comment by kennytm** [comment]
```
Building stage2 tool cargo (x86_64-pc-windows-msvc)
    Updating crates.io index
error: failed to select a version for `openssl`.
    ... required by package `rustc-workspace-hack v1.0.0 (C:\projects\rust\src\tools\rustc-workspace-hack)`
    ... which is depended on by `cargo v0.32.0 (C:\projects\rust\src\tools\cargo)`
versions that meet the requirements `^0.10.12` are: 0.10.12
all possible versions conflict with previously selected packages.
  previously selected package `openssl v0.10.11`
    ... which is depended on by `crypto-hash v0.3.1`
    ... which is depended on by `cargo v0.32.0 (C:\projects\rust\src\tools\cargo)`
failed to select a version for `openssl` which could resolve this conflict
```

**Comment by kennytm** [comment]
`winapi` additionally needs the `sspi` feature.

<details>

```
duplicate artfacts found when compiling a tool, this typically means that something was recompiled because a transitive dependency has different features activated than in a previous build:
the following dependencies are duplicated although they have the same features enabled:
  backtrace 0.3.9 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` ("C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libbacktrace-a98b5133c4ddb86f.rlib")
    `cargo` ("C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libbacktrace-c16c8fefdc222c33.rlib")
the following dependencies have different features:
  winapi 0.3.6 (registry+https://github.com/rust-lang/crates.io-index)
    `clippy-driver` additionally enabled features {} at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libwinapi-47ff9f61614f583f.rlib"
    `cargo` additionally enabled features {"sspi"} at "C:\\projects\\rust\\build\\x86_64-pc-windows-msvc\\stage2-tools\\x86_64-pc-windows-msvc\\release\\deps\\libwinapi-75b667478104b90c.rlib"
to fix this you will probably want to edit the local src/tools/rustc-workspace-hack/Cargo.toml crate, as that will update the dependency graph to ensure that these crates all share the same feature set
```

</details>

**Comment by alexcrichton** [comment]
@nrc so that [last failure](https://ci.appveyor.com/project/rust-lang/rust/builds/19587869/job/u8l3v8ubxhepd9ri) is because the `rustfmt` crate doesn't already depend on `rustc-workspace-hack` in this repository. I've now [added that to `rustfmt`](https://github.com/rust-lang-nursery/rustfmt/commit/3d07646858698a1d904c63d3ccc4d9c041928ed8) but wanted to inquire about updating rustfmt, is that something which can be done on a one-off basis or is largely sticking to tags now?

**Comment by nrc** [comment]
We trying to update on tags only now. I'm working on a PR I want to get out today. Either way I'll make a new version tonight.

**Comment by alexcrichton** [comment]
Ok sure thing! It's just my luck that whenever I try to update Cargo it hits the period of "well now we can't break rls/rustfmt"

**Comment by nrc** [comment]
I've updated Rustfmt and RLS, If you take Rustfmt at tag 0.99.6 and the latest RLS, it should work. Let me know if there's anything I need to fixup.

**Comment by alexcrichton** [comment]
Thanks @nrc! It looks like racer may also need an update to avoid duplicating `rustc-ap-syntax`, do you know if that's in the works?

**Comment by nrc** [comment]
I can do that this morning

**Comment by nrc** [comment]
hmm, updating Racer was harder than I thought - currently blocked on https://github.com/racer-rust/racer/issues/974

**Comment by Xanewok** [comment]
@nrc @alexcrichton with https://github.com/rust-lang-nursery/rls/pull/1095 merged, the `rustc-ap-*` are deduplicated in the RLS (currently `274.0.0` is used)

**Comment by alexcrichton** [comment]
@Xanewok thanks so much for the update!

@bors: r=Mark-Simulacrum

**Comment by kennytm** [comment]
@bors retry

"The system is going down for power off NOW!" again?!

**Pull request merged**